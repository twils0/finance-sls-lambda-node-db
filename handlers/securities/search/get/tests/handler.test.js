const { loadPool } = require('../../../../../functions/db/loadPool');
const { searchSecurities } = require('../../../../../functions/db/searchSecurities');

const { get } = require('../handler');

global.console.log = jest.fn();
jest.mock('../../../../../functions/db/loadPool', () => ({
  loadPool: jest.fn(),
}));
jest.mock('../../../../../functions/db/searchSecurities', () => ({
  searchSecurities: jest.fn(),
}));
jest.mock('../../../../../functions/errorResponse', () => ({
  errorResponse: jest.fn(error => error),
}));

const event = {
  query: {
    search: 'testSearch',
  },
};
global.pool = { test: 'testPool' };
const securities = [{ test1: 'testSecurity1' }, { test2: 'testSecurity2' }];
const response = {
  status: 200,
  body: {
    securities,
  },
};
const emptyResponse = {
  status: 200,
  body: {
    securities: null,
  },
};
const callback = jest.fn();

describe('handlers', () => {
  describe('securities', () => {
    describe('get', () => {
      afterEach(() => {
        loadPool.mockReset();
        searchSecurities.mockReset();
        global.console.log.mockReset();
        callback.mockReset();
      });

      it('fails and returns error, when loadPool throws error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        loadPool.mockReturnValue(Promise.reject(error));

        await get(event, null, callback);

        expect(loadPool).toBeCalled();
        expect(callback).toBeCalledWith(error, emptyResponse);
        expect(global.console.log).toMatchSnapshot();
      });

      it("fails and returns error, when missing 'search' in query", async () => {
        const wrongEvent = {
          query: {},
        };
        loadPool.mockReturnValue(Promise.resolve());
        callback.mockImplementation((error, resp) => {
          expect(error).toMatchSnapshot();
          expect(resp).toEqual(emptyResponse);
        });

        await get(wrongEvent, null, callback);

        expect(loadPool).toBeCalled();
        expect(callback).toBeCalled();
        expect(global.console.log).toMatchSnapshot();
      });

      it('fails and returns error, when searchSecurities returns an error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        loadPool.mockReturnValue(Promise.resolve());
        searchSecurities.mockReturnValue(Promise.reject(error));

        await get(event, null, callback);

        expect(loadPool).toBeCalled();
        expect(searchSecurities).toBeCalledWith(event.query);
        expect(callback).toBeCalledWith(error, emptyResponse);
        expect(global.console.log).toMatchSnapshot();
      });

      it('correctly responds with a security', async () => {
        loadPool.mockReturnValue(Promise.resolve());
        searchSecurities.mockReturnValue(Promise.resolve({ securities }));

        await get(event, null, callback);

        expect(loadPool).toBeCalled();
        expect(searchSecurities).toBeCalledWith(event.query);
        expect(callback).toBeCalledWith(null, response);
        expect(global.console.log).toMatchSnapshot();
      });
    });
  });
});
