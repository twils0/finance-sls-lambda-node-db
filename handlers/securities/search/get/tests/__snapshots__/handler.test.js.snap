// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlers securities get correctly responds with a security 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
    Array [
      "searchSecurities
",
      Object {
        "search": "testSearch",
      },
      "
",
    ],
    Array [
      "searchSecurities - success
",
      Array [
        Object {
          "test1": "testSecurity1",
        },
        Object {
          "test2": "testSecurity2",
        },
      ],
      "
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers securities get fails and returns error, when loadPool throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers securities get fails and returns error, when missing 'search' in query 1`] = `
Object {
  "code": "invalid_query",
  "message": "Please provide a 'search' key as a query string",
}
`;

exports[`handlers securities get fails and returns error, when missing 'search' in query 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers securities get fails and returns error, when searchSecurities returns an error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
    Array [
      "searchSecurities
",
      Object {
        "search": "testSearch",
      },
      "
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers securities get fails and returns error, when testAuth throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
